src/event_timer.py

```diff
@@ -0,0 +1,122 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""EventTimer for scheduling dispatch of juju event on regular intervals."""
+
+import subprocess  # nosec B404
+from pathlib import Path
+from typing import Optional, TypedDict
+
+from jinja2 import Environment, FileSystemLoader
+
+
+class TimerEnableError(Exception):
+    """Raised when unable to enable a event timer."""
+
+
+class TimerDisableError(Exception):
+    """Raised when unable to disable a event timer."""
+
+
+class EventConfig(TypedDict):
+    """Configuration used by service and timer templates."""
+
+    event: str
+    interval: float
+    jitter: float
+    timeout: float
+    unit: str
+
+
+class EventTimer:
+    """Manages the timer to emit juju events at regular intervals.
+
+    Attributes:
+        unit_name (str): Name of the juju unit to emit events to.
+    """
+
+    _systemd_path = Path("/etc/systemd/system")
+
+    def __init__(self, unit_name: str):
+        """Construct the timer manager.
+
+        Args:
+            unit_name: Name of the juju unit to emit events to.
+        """
+        self.unit_name = unit_name
+        self._jinja = Environment(loader=FileSystemLoader("templates"), autoescape=True)
+
+    def _render_event_template(self, template_type: str, event_name: str, context: EventConfig):
+        """Write event configuration files to systemd path.
+
+        Args:
+            template_type: Name of the template type to use. Can be 'service' or 'timer'.
+            event_name: Name of the event to schedule.
+            context: Addition configuration for the event to schedule.
+        """
+        template = self._jinja.get_template(f"dispatch-event.{template_type}.j2")
+        dest = self._systemd_path / f"ghro.{event_name}.{template_type}"
+        dest.write_text(template.render(context))
+
+    def ensure_event_timer(
+        self, event_name: str, interval: float, timeout: Optional[float] = None
+    ):
+        """Ensure that a systemd service and timer are registered to dispatch the given event.
+
+        The interval is how frequently, in minutes, that the event should be dispatched.
+
+        The timeout is the number of seconds before an event is timed out. If not given or 0,
+        it defaults to half the interval period.
+
+        Args:
+            event_name: Name of the juju event to schedule.
+            interval: Number of minutes between emitting each event.
+
+        Raises:
+            TimerEnableError: Timer cannot be started. Events will be not emitted.
+        """
+        context: EventConfig = {
+            "event": event_name,
+            "interval": interval,
+            "jitter": interval / 4,
+            "timeout": timeout or (interval * 30),
```

Comment:
How does this default to half the interval period as the docstring says?

Reply:
The interval is in minutes and the timeout is seconds
So if you take interval * 30 it's equivalent to (interval * 60)/2 so half the interval

Reply:
Makes sense, but it's not abvious from the code. I'd go with `"timeout": timeout or (interval * 60 / 2),`

Reply:
Minor comments